common/types/yamlUtils.go

```diff
@@ -19,25 +21,114 @@ func ReadYaml(filePath string, primaryKey string, content interface{}, hooks []m
 	// handle proto enums
 	opts := []viper.DecoderConfigOption{
 		func(dc *mapstructure.DecoderConfig) {
-			dc.ErrorUnset = true
 			dc.ErrorUnused = true
 		},
 	}
 	if len(hooks) != 0 {
+		hookFuncs := make([]mapstructure.DecodeHookFunc, 0, len(hooks))
 		for _, h := range hooks {
-			opts = append(opts, viper.DecodeHook(h))
+			hookFuncs = append(hookFuncs, mapstructure.ComposeDecodeHookFunc(h))
 		}
+		opts = append(opts, viper.DecodeHook(mapstructure.ComposeDecodeHookFunc(hookFuncs...)))
 	}
 
-	err := viper.ReadInConfig()
+	err = viper.ReadInConfig()
 	if err != nil {
-		return err
+		return missingFields, err
 	}
 
 	err = viper.GetViper().UnmarshalKey(primaryKey, content, opts...)
 	if err != nil {
-		return err
+		return missingFields, err
 	}
 
-	return nil
+	missingFields = findMissingFields(content)
+	return missingFields, nil
+}
+
+func findMissingFields(content interface{}) []string {
+	var missingFields []string
+
+	expectedFields := make(map[string]bool)
+	contentValue := reflect.ValueOf(content)
+	contentType := contentValue.Type()
+
+	// Ensure content is a struct or a pointer to a struct
+	if contentType.Kind() != reflect.Struct && (contentType.Kind() != reflect.Ptr || contentType.Elem().Kind() != reflect.Struct) {
+		return missingFields
+	}
+
+	// Get the type of the struct
+	if contentType.Kind() == reflect.Ptr {
+		contentType = contentType.Elem()
+		contentValue = contentValue.Elem()
+	}
+
+	// Extract the expected field names from the struct
+	for i := 0; i < contentType.NumField(); i++ {
+		field := contentType.Field(i)
+		fieldName := field.Name
+		expectedFields[fieldName] = true
+	}
+
+	// Check if each expected field is present in the unmarshaled content
+	for fieldName := range expectedFields {
+		fieldValue := contentValue.FieldByName(fieldName)
+		if !fieldValue.IsValid() || fieldValue.IsZero() {
```

Comment:
This can be tricky: what if the fieldValue is legitimately zero, in the input?

Reply:
I'll open a tech debt ticket

Reply:
specifically, the problem is that this can't distinguish an implicit zero value (due to missing field) from an explicit zero value. Therefore, if a specific field:
* _may_ have the zero value, -and-
* has its default value set to non-zero value
then it would be impossible to set the field in the yaml config file explicitly to zero value, because it will always be interpreted as missing (since the code assumes "missing if-and-only-if zero"), and hence always be overwritten with the default (non-zero) value.

to solve this we need a reliable method to find really missing fields (as opposed to present-zero-values). one way to do that is to use viper to also read the config into a map (of things). then, while scanning through the struct looking into its fields, replace the current test with a test that looks up the field in the map - if not found there then it was missing from the yaml config file.